src/table/iterator.rs

```diff
@@ -0,0 +1,180 @@
+use super::builder::{Header, HEADER_SIZE};
+use super::Block;
+use crate::util::{self, KeyComparator, COMPARATOR};
+use bytes::{Bytes, BytesMut};
+use std::sync::Arc;
+
+/// Errors that may encounter during iterator operation
+#[derive(Clone, Debug)]
+pub enum IteratorError {
+    EOF,
+    // TODO: As we need to clone Error from block iterator to table iterator,
+    // we had to save `crate::Error` as String. In the future, we could let all
+    // seek-related function to return a `Result<()>` instead of saving an
+    // error inside struct.
+    Error(String),
+}
+
+impl IteratorError {
+    /// Check if iterator has reached its end
+    pub fn is_eof(&self) -> bool {
+        matches!(self, IteratorError::EOF)
+    }
+}
+
+enum SeekPos {
+    Origin,
+    Current,
+}
+
+/// Block iterator iterates on an SST block
+// TODO: support custom comparator
+struct BlockIterator {
+    /// current index of iterator
+    idx: isize,
+    /// base key of the block
+    base_key: Bytes,
+    /// key of current entry
+    key: BytesMut,
+    /// raw value of current entry
+    val: Bytes,
+    /// block data in bytes
+    data: Bytes,
+    /// block struct
+    // TODO: use `&'a Block` if possible
+    block: Arc<Block>,
+    /// previous overlap key, used to construct key of current entry from
+    /// previous one faster
+    perv_overlap: u16,
+    /// iterator error in last operation
+    err: Option<IteratorError>,
+}
+
+impl BlockIterator {
+    pub fn new(block: Arc<Block>) -> Self {
+        let data = block.data.slice(..block.entries_index_start);
+        Self {
+            block,
+            err: None,
+            base_key: Bytes::new(),
+            key: BytesMut::new(),
+            val: Bytes::new(),
+            data,
+            perv_overlap: 0,
+            idx: 0,
+        }
+    }
+
+    #[inline]
+    fn entry_offsets(&self) -> &[u32] {
+        &self.block.entry_offsets
+    }
+
+    fn set_idx(&mut self, i: isize) {
```

Comment:
Can it use usize?

Reply:
badger will invalidate the iterator when index is negative. We could use `usize` in parameter, but will have to store `isize` inside the iterator.

Reply:
Hmm... I think it's better to use `isize` here, as `self.idx` should always contain the same value as input parameter. When user provides a `usize` larger than `isize` max, there could be inconsistency. So I would prefer using `isize` in all iterators.

Reply:
Why not make `self.idx` as `usize`?

Reply:
I've refactored iterator to store `usize`, and added extra check for reaching reverse iterator end in `seek_last`.